### ISSUE TITLE
Bug with ResizingImageProcessor and .aspectFill (possibly .aspectFit as well)

### ISSUE THREAD

**Issue by bmetitiri** [onevcat/Kingfisher#645]
The image returned by `ResizingImageProcessor(targetSize: <any square size>, contentMode: .aspectFill)` does not respect the aspect ratio of the input image, and has a size that does not match the `targetSize`.

Specifically I am loading the URL https://scontent.xx.fbcdn.net/v/t1.0-1/p200x200/1511435_10153809629850504_646477087_n.jpg?oh=8dc399c391d96f2a6902f02979708557&oe=5905D119 which is much taller than it is wide. The visual result is:

<img width="106" alt="screen shot 2017-04-04 at 5 10 12 pm" src="https://cloud.githubusercontent.com/assets/14278459/24684295/bf4870e0-1959-11e7-8b4b-ab67511494bb.png">

which is clearly squashed. Additionally, I am also applying the `RoundCornerImageProcessor` to make the image a circle, but when passing in a `targetSize` of 64x64 to the `ResizingImageProcessor`, the returned image has a size of 64x99, so when passed into `RoundCornerImageProcessor` without a target size the rounding is applied incorrectly. Passing my actual target size to `RoundCornerImageProcessor` as well fixes this, but it doesn't seem like it should be necessary.

TLDR: 
```
let targetSize = CGSize(width: 64, height: 64)
let resize = ResizingImageProcessor(targetSize: targetSize, contentMode: .aspectFill)
let round = RoundCornerImageProcessor(cornerRadius: targetSize.height/2)
(resize >> round).process(<stuff>)
```
should return a non-squashed image in a circle no matter what image you pass in, but it doesn't.

**Comment by onevcat** [comment]
Hi,

The behavior of `ResizingImageProcessor` is intended and correct here. The `target` size is only a "target", or say, a reference size when you are passing a `contentMode`. It could be not the same as the final image size since we have to respect the ratio if there is an "aspect" mode. 

Here you are using `. aspectFill`, and the resized image has to be 64x99 to respect the aspect ratio of your image. And I believe the final image size should be also 64x99, without squashed, it might be something like this (if you were using a `RoundCornerImageProcessor` with 32.0 as its `cornerRadius `):

![snip20170405_1](https://cloud.githubusercontent.com/assets/1019875/24686094/98ba3bbe-19ec-11e7-842a-5d1fd82febdd.png)

Maybe you are using a "Scale To Fill" as your image view content mode, so you are getting a squashed image. You could try to change it to something like "Center" to see whether it could display the final image.

We should improve the documentation for `targetSize` of `ResizingImageProcessor` to make it clearer, however, it is the intended way that the resizing processor working.

For your case, I suggest you to use a `CroppingImageProcessor` (or a combination of `CroppingImageProcessor` and `ResizingImageProcessor`) to get a squared version of your image, and then `RoundCornerImageProcessor` to get a proper roundcornered image.

**Comment by bmetitiri** [comment]
Oh wow, I was completely misinterpreting how `ResizingImageProcessor` is supposed to work apparently. I assumed you passed in the target size for the resized image that you want, and I assumed that the addition of the `.aspectFill` mode would cause it to crop to reach the target size rather than squashing. 

Maybe if you added a note to the `contentMode` docs saying that it overrides the `targetSize`, and that you should see `CroppingImageProcessor` for cropping? I guess it should seem obvious that you use `CroppingImageProcessor` for cropping, but I was just super assuming the `.aspectFill` mode would act like that mode on a `UIImageView`.

**Comment by onevcat** [comment]
@bmetitiri Yes, you are right. I will update the documentation for it soon. And maybe also rename the parameter labels.

**Comment by onevcat** [comment]
https://github.com/onevcat/Kingfisher/pull/646

**Comment by bmetitiri** [comment]
So a follow up question, how do I get `CroppingImageProcessor` to respect the scale factor? It seems to be behaving differently with this processor than with the others. I changed my custom processor code to:

```
func process(item: ImageProcessItem, options: KingfisherOptionsInfo) -> Image? {
  let resize = ResizingImageProcessor(targetSize: targetSize, contentMode: .aspectFill)
  let crop = CroppingImageProcessor(size: targetSize)
  let round = RoundCornerImageProcessor(cornerRadius: targetSize.height / 2)
  var newOptions = options
  newOptions.append(.scaleFactor(UIScreen.main.scale))
  return ((resize >> crop) >> round).process(<stuff>, options: newOptions)
}
```

but the image that comes out after the crop step is half the width and height of the `targetSize`, cropping out 3/4 of the image. If I double my target size and don't pass in a scale factor, it works as expected, but this behavior just seems strange.

**Comment by onevcat** [comment]
Oops, yes. We need consider image scale when cropping it. I'll give it a fix soon.

However, I suggest to pass the `.scaleFactor` from the image setting methods instead of injecting it in the processor. You will lose it when you trying to load it from disk cache, so the image might be at scale of 1 at there.

**Comment by onevcat** [comment]
Scaling fixed in #649
